### PR TITLE
VFB-5 Data Input Components

### PR DESCRIPTION
**Pull request by daniel-conlan-softwire**
Implemented input components for use across the application including:
- Free Form Text Input
- Dropdown List
- Radio Groups
- Checkbox Groups

Tests have been included for each new component, and change handler factory functions have been provided for ease of use.

====================

- [X] Make sure you've verified it works via `npm run dev`
- [X] Make sure you've linted via `npm run lint`
    - can run `npm run fix` to try and fix as any mistakes automatically as possible!
- [X] Make sure you've tested via `npm run test`

### CODE REVIEW

File: src/components/DataInput/CheckboxInput.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    groupLabel?: string;
+    onChange: (event: React.ChangeEvent<HTMLInputElement>) => void; // eslint-disable-line no-unused-vars
+}
+
+const CheckboxInput: React.FC<Props> = (props) => {
+    return (
+        <FormControl>
+            <FormLabel>{props.groupLabel}</FormLabel>
+            <FormGroup>
+                {props.optionLabels.map((label, index) => {
```
Comment (Var1377) [inline src/components/DataInput/CheckboxInput.tsx:18]: For a checkbox component we probably want just a single checkbox

File: src/components/DataInput/CheckboxInput.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
+"use client";
+
+import React from "react";
+import { Checkbox, FormControl, FormControlLabel, FormGroup, FormLabel } from "@mui/material";
+
+interface Props {
+    optionLabels: string[];
+    optionKeys: string[];
+    groupLabel?: string;
```
Comment (Var1377) [inline src/components/DataInput/CheckboxInput.tsx:9]: For a label you could use props.children to add flexibility and clarity in downstream markup

File: src/components/DataInput/DropdownListInput.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
+"use client";
+
+import React, { ReactNode } from "react";
+import { MenuItem, InputLabel, Select, FormControl, SelectChangeEvent } from "@mui/material";
+
+interface Props {
+    optionLabels: string[];
+    optionValues: string[];
+    listLabel?: string;
```
Comment (Var1377) [inline src/components/DataInput/DropdownListInput.tsx:9]: props.children

File: src/components/DataInput/DropdownListInput.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
+"use client";
+
+import React, { ReactNode } from "react";
+import { MenuItem, InputLabel, Select, FormControl, SelectChangeEvent } from "@mui/material";
+
+interface Props {
+    optionLabels: string[];
+    optionValues: string[];
```
Comment (Var1377) [inline src/components/DataInput/DropdownListInput.tsx:8]: for this sort of API, a [string,string][] type for options probably makes more sense as `optionLabels` and `optionValues` have a one-to-one correspondence

File: src/components/DataInput/changeHandlerFactories.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+            [event.target.name]: event.target.checked,
+        });
+    };
+};
+
+export const handleChangeFactory = (onChange: any): any => {
+    return (event: any) => {
+        onChange((event.target as HTMLInputElement).value);
+    };
+};
```
Comment (Var1377) [inline src/components/DataInput/changeHandlerFactories.ts:26]: this entire file seems redundant, you can use the native event types instead

Comment (daniel-conlan-softwire) [reply src/components/DataInput/changeHandlerFactories.ts:26]: Yeah, we were just trying to make things easier for you guys. Turns out you're too advanced for our niceties:(

I shall remove the file for you

**Review by TomCusack-Softwire — CHANGES_REQUESTED** [review]

File: src/components/DataInput/DataInput.cy.tsx [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+                ]}
+                onChange={getRadioGroupHandler(() => console.log("Radio Changed"))}
+            />
+        );
+    });
+
+    describe("Change Handlers", () => {
+        it("Change handler for checkbox works", () => {
+            const onChangeSpy = cy.spy().as("onChangeSpy");
+            const unwrapEvent = (e: any): void => {
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/components/DataInput/DataInput.cy.tsx:87]: Done! Added types to the events

File: src/components/DataInput/DataInput.cy.tsx [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+                        ["B", "b"],
+                        ["C", "c"],
+                    ]}
+                    listTitle="Dropdown List"
+                    defaultValue="Default"
+                    onChange={unwrapEvent}
+                />
+            );
+
+            cy.get("div[class^='MuiSelect'").click();
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/components/DataInput/DataInput.cy.tsx:119]: Done! Added the closing ]

File: src/components/DataInput/CheckboxInput.tsx [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import { Checkbox, FormControl, FormControlLabel } from "@mui/material";
+
+interface Props {
+    label?: string;
+    onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
+}
+
+const CheckboxInput: React.FC<Props> = (props) => {
+    return (
+        <FormControl>
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/components/DataInput/CheckboxInput.tsx:13]: FormControl has the advantage over a div or a fragment in that it allows us to customise the inside of it later, with inbuilt functionality such as error messages, hiding the label and changing the colour of the element.

Basically, it's a wrapper component that allows state to be shared amongst each internal element of the form.

FormControlLabel is just another component that groups together all of the individual pieces of an input component together.

File: src/components/DataInput/DataInput.cy.tsx [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+                    ["B", "b"],
+                    ["C", "c"],
+                ]}
+                groupTitle="Radio Group"
+                onChange={getRadioGroupHandler(() => console.log("Radio Changed"))}
+            />
+        );
+    });
+
+    it("renders without optional props", () => {
```
[the remark this answers is not in the document]
Comment (TomCusack-Softwire) [reply src/components/DataInput/DataInput.cy.tsx:54]: can the tests reflect this? remove all the non-optional props and see if it still renders etc

File: src/components/DataInput/README.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+For example, in changing a state variable `bool` using a `<CheckboxInput />` component, the event handler `getCheckboxHandler` may be used in the following manner:
+
+```typescript jsx
+const [bool, setBool] = useState(false);
+```
+
+```typescript jsx
+<CheckboxInput label="Option 1" onChange={getCheckboxHandler(setBool)} />
+````
+\
```
Comment (TomCusack-Softwire) [inline src/components/DataInput/README.md:31]: \ potentially left from a previous commit

File: src/components/DataInput/inputHandlerFactories.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+export const getDropdownListHandler = (setValue: valueSetter): selectChangeEventHandler => {
+    return (event: SelectChangeEvent<unknown>) => {
+        setValue(event.target.value);
+    };
+};
+
+export const getCheckboxHandler = (setBoolean: booleanSetter): changeEventHandler => {
+    return (event: React.ChangeEvent<HTMLInputElement>) => {
+        setBoolean((event.target as HTMLInputElement).checked);
```
Comment (TomCusack-Softwire) [inline src/components/DataInput/inputHandlerFactories.ts:42]: does the `as` need to be here? I assume the `ChangeEvent<HTMLInputElement>` implies this

File: src/components/DataInput/inputHandlerFactories.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
+import React from "react";
+import { SelectChangeEvent } from "@mui/material";
+
+type valueSetter = {
+    (value: any): void;
```
Comment (TomCusack-Softwire) [inline src/components/DataInput/inputHandlerFactories.ts:5]: is this 'any' not just a `string`? afaik all the inputs you have are string valued

**Review by TomCusack-Softwire — CHANGES_REQUESTED** [review]
Almost there! Just a few minor comments 🙂

File: src/components/DataInput/README.md (hunk 10 above)
Comment (daniel-conlan-softwire) [reply src/components/DataInput/README.md:31]: 'Tis a newline

File: src/components/DataInput/DataInput.cy.tsx [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+            cy.get("li[data-value='b']").click();
+            cy.get("@onChangeSpy").should("have.been.calledWith", "b");
+
+            cy.get("li[data-value='c']").click();
+            cy.get("@onChangeSpy").should("have.been.calledWith", "c");
+        });
+
+        it("Change handler for freeform text works", () => {
+            const onChangeSpy = cy.spy().as("onChangeSpy");
+            const unwrapEvent = (e: React.ChangeEvent<HTMLInputElement>): void => {
```
Comment (TomCusack-Softwire) [inline src/components/DataInput/DataInput.cy.tsx:126]: change `e:` to `event:` everywhere and I'm happy :)

**Review by TomCusack-Softwire — APPROVED** [review]

File: package.json [hunk 14]
```diff
@@ -26,10 +26,14 @@
   },
   "dependencies": {
     "@cypress/code-coverage": "^3.10.7",
+    "@emotion/react": "^11.11.1",
+    "@emotion/styled": "^11.11.0",
+    "@fontsource/roboto": "^5.0.5",
```
Comment (SamYap-Softwire) [inline package.json:31]: is this used in the code?

File: src/components/DataInput/DataInput.cy.tsx [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+    getRadioGroupHandler,
+} from "@/components/DataInput/inputHandlerFactories";
+import { SelectChangeEvent } from "@mui/material";
+
+describe("Data Input Components", () => {
+    it("renders", () => {
+        cy.mount(
+            <CheckboxInput
+                label="LABEL"
+                onChange={getCheckboxHandler(() => console.log("Checkbox Changed"))}
```
Comment (SamYap-Softwire) [inline src/components/DataInput/DataInput.cy.tsx:19]: Is there a reason why we are using console.log? Does it help with the test?

File: src/components/DataInput/DataInput.cy.tsx [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+            />
+        );
+        cy.mount(
+            <DropdownListInput
+                labelsAndValues={[
```
Comment (SamYap-Softwire) [inline src/components/DataInput/DataInput.cy.tsx:24]: I think this has been used in many of the individual tests. Would it be better to make it a const before all the tests?

**Review by SamYap-Softwire — COMMENTED** [review]
I will continue the review on Monday. (Currently done with DataInput.cy.tsx)

File: src/components/DataInput/FreeFormTextInput.tsx
Comment (paopaosoftwire) [inline src/components/DataInput/FreeFormTextInput.tsx:1]: No need to change anything but I just want to know whether using FreeFormTextInput is equivalent to using TextField from material UI directly. Can I provide the TextField with getFreeFormTextHandler?

**Comment by paopaosoftwire** [comment]
Everything looks pretty clean to me!

**Pull request merged**